runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java

```diff
@@ -0,0 +1,245 @@
+				uiOption.setBorder(BorderFactory.createEmptyBorder(0, 8, 0, 0));
+
+				// Adjust XP bonus depending on check-state of the boxes.
+				uiCheckbox.addActionListener(e -> adjustXPBonus(uiCheckbox.isSelected(), bonus.value));
+
+				uiOption.add(uiLabel, BorderLayout.WEST);
+				uiOption.add(uiCheckbox, BorderLayout.EAST);
+				add(uiOption);
+			}
+		}
+	}
+
+	private void renderActionSlots()
+	{
+		// Wipe the list of references to the slot components.
+		uiActionSlots.clear();
+
+		// Create new components for the action slots.
+		for (SkillDataEntry action : skillData.actions)
+		{
+			UIActionSlot slot = new UIActionSlot(action);
+			uiActionSlots.add(slot); // Keep our own reference.
+			add(slot); // Add component to the panel.
+		}
+
+		// Clear the icon cache once we're done making slots.
+		cacheItemIcon.clear();
+
+		// Refresh the rendering of this panel.
+		revalidate();
+		repaint();
+	}
+
+	private void calculate()
+	{
+		for (UIActionSlot slot : uiActionSlots)
+		{
+			int actionCount = 0;
+			int neededXP = targetXP - currentXP;
+			double xp = slot.action.xp * xpFactor;
+
+			if (neededXP > 0)
+				actionCount = (int) Math.ceil(neededXP / xp);
+
+			slot.setText(XP_FORMAT.format(xp) + "xp - " + NumberFormat.getIntegerInstance().format(actionCount) + (actionCount > 1 ? " actions" : " action"));
+
+			if (currentLevel >= slot.action.level)
+				slot.setSlotColor(Color.GREEN);
+			else
+				slot.setSlotColor(Color.RED);
+		}
+	}
+
+	private void updateInputFields()
+	{
+		if (targetXP < currentXP)
+		{
+			targetLevel = EnforceSkillBounds(currentLevel + 1);
+			targetXP = SkillLevelToXP(targetLevel);
+		}
+
+		uiInput.setCurrentLevelInput(currentLevel);
+		uiInput.setCurrentXPInput(currentXP);
+		uiInput.setTargetLevelInput(targetLevel);
+		uiInput.setTargetXPInput(targetXP);
+		calculate();
+	}
+
+	private void adjustXPBonus(boolean addBonus, float value)
+	{
+		xpFactor += addBonus ? value : -value;
+		calculate();
+	}
+
+	private void onFieldCurrentLevelUpdated()
+	{
+		currentLevel = EnforceSkillBounds(uiInput.getCurrentLevelInput());
+		currentXP = SkillLevelToXP(currentLevel);
+		updateInputFields();
+	}
+
+	private void onFieldCurrentXPUpdated()
+	{
+		currentXP = EnforceXPBounds(uiInput.getCurrentXPInput());
+		currentLevel = XPToSkillLevel(currentXP);
+		updateInputFields();
+	}
+
+	private void onFieldTargetLevelUpdated()
+	{
+		targetLevel = EnforceSkillBounds(uiInput.getTargetLevelInput());
+		targetXP = SkillLevelToXP(targetLevel);
+		updateInputFields();
+	}
+
+	private void onFieldTargetXPUpdated()
+	{
+		targetXP = EnforceXPBounds(uiInput.getTargetXPInput());
+		targetLevel = XPToSkillLevel(targetXP);
+		updateInputFields();
+	}
+
+	private static int EnforceSkillBounds(int input)
+	{
+		return Math.min(MAX_LEVEL, Math.max(MIN_LEVEL, input));
+	}
+
+	private static int EnforceXPBounds(int input)
+	{
+		return Math.min(MAX_XP, Math.max(0, input));
+	}
+
+	private static int SkillLevelToXP(int level)
```

Comment:
I think we've got code for this in the `Experience` class in `net.runelite.api`. If we're missing functionality, it'd probably be best to add it there.